crypto/context.c

```diff
@@ -172,16 +194,24 @@ void *openssl_ctx_get_data(OPENSSL_CTX *ctx, int index)
 {
     void *data = NULL;
 
+#ifndef FIPS_MODE
     if (ctx == NULL) {
         if (!RUN_ONCE(&default_context_init, do_default_context_init))
             return NULL;
         ctx = default_context;
     }
+#endif
     if (ctx == NULL)
         return NULL;
 
     CRYPTO_THREAD_read_lock(ctx->lock);
 
+    if (ctx->dyn_indexes[index] == -1
+            && !openssl_ctx_init_index(ctx, index, NULL)) {
+        CRYPTO_THREAD_unlock(ctx->lock);
+        return NULL;
+    }
+
```

Comment:
Ah, I see what you're trying to do.  I think you're trying to make the code too smart.  If the index is -1, this should simply be an error.  (I actually let that mean that `CRYPTO_alloc_ex_data` crash, but returning NULL is fine too)

Reply:
I think this avoids *lots* of complexity. It actually means in most cases you don't need run once init code *at all*.

Reply:
Have you noticed the times when outside people have asked "if I want to add a sub-system, what things must I touch", and it turns out that they must go touch a number of things, and it's way too easy to miss something?  This recreates exactly that situation.  I was trying to make the slot allocation in the library context fairly self contained within each sub-system that does it, and the code changes I see now go exactly in the opposite direction.

Reply:
If you don't like that then we can go with the "pass the method as part of the openssl_ctx_get_data" call idea. Actually we could make a macro for each file we do that (if there are multiple of them), which knows the index and method for that file already. So something like

````
#define FOO_get_openssl_ctx_data(ctx)      openssl_ctx_get_data(ctx, FOO_INDEX, foo_ctx_method)
````

Reply:
That makes me cringe a little less.  At least, that preserves the more decentralized creation of the index.  So, uhmm, yes please (not sure I see the necessity for the macro, but [shrug])